src/finalfusion/vocab/subword.py

```diff
@@ -0,0 +1,257 @@
+
+    @property
+    def max_n(self) -> int:
+        """
+        Get the upper bound of the range of extracted n-grams.
+
+        Returns
+        -------
+        max_n : int
+            upper bound of n-gram range.
+        """
+        return self.subword_indexer.max_n
+
+    @property
+    @abstractmethod
+    def subword_indexer(
+            self
+    ) -> Union[ExplicitIndexer, FinalfusionHashIndexer, FastTextIndexer]:
+        """
+        Get this vocab's subword Indexer.
+
+        The subword indexer produces indices for n-grams.
+
+        In case of bucket vocabularies, this is a hash-based indexer
+        (:class:`.FinalfusionHashIndexer`, :class:`.FastTextIndexer`). For explicit subword
+        vocabularies, this is an :class:`.ExplicitIndexer`.
+
+        Returns
+        -------
+        subword_indexer : ExplicitIndexer, FinalfusionHashIndexer, FastTextIndexer
+            The subword indexer of the vocabulary.
+        """
+    def subwords(self, item: str, bracket: bool = True) -> List[str]:
+        """
+        Get the n-grams of the given item as a list.
+
+        The n-gram range is determined by the `min_n` and `max_n` values.
+
+        Parameters
+        ----------
+        item : str
+            The query item to extract n-grams from.
+        bracket : bool
+            Toggles bracketing the item with '<' and '>' before extraction.
+
+        Returns
+        -------
+        ngrams : List[str]
+            List of n-grams.
+        """
+        return ngrams(item, self.min_n, self.max_n, bracket)
+
+    def subword_indices(self, item: str, bracket: bool = True, with_ngrams: bool = False)\
+            -> List[Union[int, Tuple[str, int]]]:
+        """
+        Get the subword indices for the given item.
+
+        This list does not contain the index for known items.
+
+        Parameters
+        ----------
+        item : str
+            The query item.
+        bracket : bool
+            Toggles bracketing the item with '<' and '>' before extraction.
+        with_ngrams : bool
+            Toggles returning ngrams together with indices.
+
+        Returns
+        -------
+        indices : List[Union[int, Tuple[str, int]]]
+            The list of subword indices.
+        """
+        return self.subword_indexer.subword_indices(item,
+                                                    offset=len(self.words),
+                                                    bracket=bracket,
+                                                    with_ngrams=with_ngrams)
+
+    def __getitem__(self, item: str) -> Union[int, List[int]]:
+        idx = self.word_index.get(item)
+        if idx is not None:
+            return idx
+        subwords = cast(List[int], self.subword_indices(item))
+        if subwords:
+            return subwords
+        raise KeyError(f"No indices found for {item}")
+
+    def __eq__(self, other: Any) -> bool:
+        return super(SubwordVocab, self).__eq__(other) \
```

Comment:
For hashing subword vocabs comparing the indexers is faster I'd assume. Maybe swap the order? Or defer the implementation to concrete implementations?

Reply:
Simply changing order will break comparison to `SimpleVocab`s, `Vocab.__eq__` has the type check which short-circuits if the types are different. `SubwordVocab.__eq__` accesses `subword_indexer` which raises an `AttributeError` for `SimpleVocab`.

I added an additional type check in `SubwordVocab.__eq__` and changed the order of comparison. There shouldn't be much of a difference when comparing the explicit indexers first vs. the vocabulary parts, but the improvement for hash indexers should be quite large.